android/src/main/java/com/rt2zz/reactnativecontacts/ContactsProvider.java

```diff
@@ -0,0 +1,187 @@
+package com.rt2zz.reactnativecontacts;
+
+import android.content.ContentResolver;
+import android.database.Cursor;
+import android.provider.ContactsContract;
+import android.text.TextUtils;
+
+import com.facebook.react.bridge.Arguments;
+import com.facebook.react.bridge.WritableArray;
+import com.facebook.react.bridge.WritableMap;
+
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+import static android.provider.ContactsContract.CommonDataKinds.*;
+
+public class ContactsProvider {
+
+    private final ContentResolver contentResolver;
+
+    public ContactsProvider(ContentResolver contentResolver) {
+        this.contentResolver = contentResolver;
+    }
+
+    public WritableArray getContacts() {
+        Cursor cursor = contentResolver.query(
+                ContactsContract.Data.CONTENT_URI,
+                new String[]{
+                        ContactsContract.Data.CONTACT_ID,
+                        ContactsContract.Data.MIMETYPE,
+                        ContactsContract.Contacts.DISPLAY_NAME,
+                        Contactables.PHOTO_URI,
+                        ContactsContract.Contacts.HAS_PHONE_NUMBER,
+                        StructuredName.DISPLAY_NAME,
+                        StructuredName.GIVEN_NAME,
+                        StructuredName.MIDDLE_NAME,
+                        StructuredName.FAMILY_NAME,
+                        Phone.NUMBER,
+                        Phone.TYPE,
+                        Phone.LABEL,
+                        Email.DATA,
+                        Email.ADDRESS,
+                        Email.TYPE,
+                        Email.LABEL
+                },
+                ContactsContract.Data.MIMETYPE + "=? OR " + ContactsContract.Data.MIMETYPE + "=? OR " + ContactsContract.Data.MIMETYPE + "=?",
+                new String[]{Email.CONTENT_ITEM_TYPE, Phone.CONTENT_ITEM_TYPE, StructuredName.CONTENT_ITEM_TYPE},
+                null
+        );
+
+        Map<Integer, Contact> map = new LinkedHashMap<>();
+
+        while (cursor.moveToNext()) {
+
+            int contactId = cursor.getInt(cursor.getColumnIndex(ContactsContract.Data.CONTACT_ID));
```

Comment:
I took a look at the code that gets the index, under the hood only the 1st request takes time - after that you're just looking them up from a map.  I don't think pulling them out will affect performance at all, at the cost of making the code a little messier.

Happy to pop some missing brackets in though.